### PR TITLE
ObjectBuilder Exceptions

### PR DESCRIPTION
**Pull request by Maithem**
## Overview
The ObjectBuilder::open doesn't fail when some of the required
fields are null. As a consequence the failure happens at a
later point in time where the stack trace isn't informative (i.e
fails during serialization). This patch makes the open fail
right away.

Why should this be merged: Improves debugging 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import com.google.common.reflect.TypeToken;
 
 import java.util.EnumSet;
+import java.util.Objects;
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java]: Might remove setType because it requires double casting (which is superfluous) to work.

File: runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java [hunk 2]
```diff
@@ -95,6 +96,8 @@ public T open() {
         log.info("ObjectBuilder: open Corfu stream {} id {}", streamName, streamID);
 
         try {
+            verify();
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java]: All this custom code can be replaced by lombok builder. If you want to clean up this code I believe it had better use a Lombok builder. 
And also would be great if you get rid of `@data` annotation and setters in the ObjectBuilder class

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java]: Ok.

File: runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        public SMRObject<T> build() {
+            if (streamID == null && streamName != null) {
+                streamID = UUID.nameUUIDFromBytes(streamName.getBytes());
+            }
+            verify();
+            return new SMRObject<>(runtime, type, streamID, streamName, serializer, option, arguments);
+        }
+
+        public T open() {
+            final SMRObject<T> smrObject = build();
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java]: Restore logging line

File: runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java (hunk 2 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java]: It took more than annotating the class, I removed some other unnecessary code. Please take a look again. 

There's some really strange usage that I'll clean up later. For example, some code expects to access the Builder's field members before the object is built, that's why the builder has getters.

Comment (xnull) [reply runtime/src/main/java/org/corfudb/runtime/view/ObjectBuilder.java]: Getters are ok, setters not desired

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 4]
```diff
@@ -185,8 +184,11 @@ public void addCustomTypeStream(UUID streamId, ObjectBuilder ob) {
      */
     public void addIndexerToCorfuTableStream(String streamName, IndexRegistry indexRegistry) {
         UUID streamId = CorfuRuntime.getStreamID(streamName);
-        ObjectBuilder ob = new ObjectBuilder(runtime).setType(CorfuTable.class)
-                .setArguments(indexRegistry).setStreamID(streamId);
+        SMRObject.Builder ob = SMRObject.builder()
+                .runtime(runtime)
+                .setType(CorfuTable.class)
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: use `type(CorfuTable)` instead of setType

File: runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            this.type = (Class<T>) typeToken.getRawType();
+            return (SMRObject.Builder<R>) this;
+        }
+
+        public SMRObject.Builder<T> setArguments(Object ... arguments) {
+            this.arguments = arguments;
+            return this;
+        }
+
+        public SMRObject.Builder<T> setStreamName(String streamName) {
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java]: This is a public API. 

Lombok gives you builder.streamName, but there's a lot of test code that uses builder.setStreamName.
Its actually not just the test code, external consumers use this API too and I dont want this PR to break it.

File: runtime/src/main/java/org/corfudb/util/serializer/JsonSerializer.java [hunk 6]
```diff
@@ -85,7 +85,7 @@ public Object deserialize(ByteBuf b, CorfuRuntime rt) {
     public void serialize(Object o, ByteBuf b) {
         String className = o == null ? "null" : o.getClass().getName();
         if (className.endsWith(ICorfuSMR.CORFUSMR_SUFFIX)) {
-            className = "CorfuObject";
+            className = "SMRObject";
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/util/serializer/JsonSerializer.java]: I am not sure I understand why this change was made. In our case, CorfuObject serves as a simple annotation. In fact, the entire `CorfuObject` interface is marked as deprecated.

File: runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    final String streamName;
+
+    @NonNull
+    final ISerializer serializer;
+
+    @NonNull
+    final ObjectOpenOption option;
+
+    @NonNull
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java]: Codacy Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 15
- Added 3
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java  4
         

Clones removed
==============
+ test/src/test/java/org/corfudb/runtime/collections/CorfuMapTest.java  -1
+ test/src/test/java/org/corfudb/runtime/collections/SMRMapTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447&bid=15577756)

File: runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    final private CorfuRuntime runtime;
+
+    @NonNull
+    final Class<T> type;
+
+    @NonNull
+    @SuppressWarnings("checkstyle:abbreviation")
+    final UUID streamID;
+
+    final String streamName;
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java]: Codacy Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 9]
```diff
@@ -171,9 +170,9 @@ public FastObjectLoader addStreamsToLoad(List<String> streamsToLoad) {
     /**
      * We can register streams with non-default type
      */
-    private final Map<UUID, ObjectBuilder> customTypeStreams = new HashMap<>();
+    private final Map<UUID, SMRObject.Builder> customTypeStreams = new HashMap<>();
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

File: runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+import org.corfudb.util.serializer.Serializers;
+
+import java.util.UUID;
+
+/**
+ * Created by mwei on 4/6/16.
+ */
+@Slf4j
+@Getter
+@Builder(builderClassName = "Builder")
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java]: Its just syntactic sugar. By default it will be named SMRObject.SMRObjectBuilder which I find a little superfluous, so instead this will generate a SMRObject.Builder instead.

File: runtime/src/main/java/org/corfudb/util/serializer/JsonSerializer.java (hunk 6 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/util/serializer/JsonSerializer.java]: Good catch, I actually didn't want to change this, intellij did it. 

In this context, it actually doesn't matter because there are problems to serializing deserializing SMRObjects, which @WenbinZhu  is removing.

File: runtime/src/main/java/org/corfudb/runtime/collections/FGMap.java [hunk 11]
```diff
@@ -1,407 +0,0 @@
...
-import org.corfudb.runtime.object.ICorfuSMR;
-import sun.misc.CRC16;
-
-/**
- * Created by mwei on 3/29/16.
- */
-@SuppressWarnings("checkstyle:abbreviation")
-@CorfuObject(constructorType = ConstructorType.PERSISTED,
-        objectType = ObjectType.STATELESS)
-public class FGMap<K, V> extends AbstractCorfuWrapper<FGMap<K,V>> implements Map<K, V>,
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/FGMap.java]: The purpose of FGMap was to build a "sharded" map from multiple SMRMaps. The idea was to have conflict resolution on a shard level, but now we have a more granular mechanism to detect conflicts (i.e. conflict keys), so this is no longer needed. Further more, there are bugs in the implementation, the shards are actually SMRObjects, but SMRObject composition is broken, some annotations dont generate the right code. The VLO also doesn't handle SMR composition. 

Long story short, it is not needed and it has bugs.

File: runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java]: Done.

**Comment by Maithem** [comment]
Addressed all comments.

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java  4
         

Clones removed
==============
+ test/src/test/java/org/corfudb/runtime/collections/SMRMapTest.java  -1
+ test/src/test/java/org/corfudb/runtime/collections/CorfuMapTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447&bid=15577756)

File: test/src/test/java/org/corfudb/runtime/object/ObjectBuilderTest.java [hunk 12]
```diff
@@ -1,9 +1,16 @@
 package org.corfudb.runtime.object;
 
+import com.google.common.reflect.TypeToken;
+import org.corfudb.protocols.wireprotocol.CorfuMsg;
```
Comment (annym) [inline test/src/test/java/org/corfudb/runtime/object/ObjectBuilderTest.java]: Codacy Issue found: [Unused import - org.corfudb.protocols.wireprotocol.CorfuMsg.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

Comment (annym) [inline test/src/test/java/org/corfudb/runtime/object/ObjectBuilderTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.corfudb.protocols.wireprotocol.CorfuMsg'](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

File: test/src/test/java/org/corfudb/runtime/object/ObjectBuilderTest.java [hunk 13]
```diff
@@ -41,6 +48,19 @@ public Example(Child base) {
         }
     }
 
+    @Test
+    public void testObjectWithInvalidStreamId() {
+        CorfuRuntime rt = new CorfuRuntime();
+        assertThatThrownBy(() -> {
+            CorfuTable<String, String> ob = SMRObject.builder()
```
Comment (annym) [inline test/src/test/java/org/corfudb/runtime/object/ObjectBuilderTest.java]: Codacy Issue found: [Avoid unused local variables such as 'ob'.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java  4
         

Clones removed
==============
+ test/src/test/java/org/corfudb/runtime/collections/SMRMapTest.java  -1
+ test/src/test/java/org/corfudb/runtime/collections/CorfuMapTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447&bid=15577756)

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 14]
```diff
@@ -170,9 +171,9 @@ public FastObjectLoader addStreamsToLoad(List<String> streamsToLoad) {
     /**
      * We can register streams with non-default type
      */
-    private final Map<UUID, ObjectBuilder> customTypeStreams = new HashMap<>();
+    private final Map<UUID, Builder> customTypeStreams = new HashMap<>();
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

**Comment by Maithem** [comment]
@vjeko @annym Let me know if you have more comments, otherwise lets get this in. Thanks.

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java  4
         

Clones removed
==============
+ test/src/test/java/org/corfudb/runtime/collections/SMRMapTest.java  -1
+ test/src/test/java/org/corfudb/runtime/collections/CorfuMapTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447&bid=15577756)

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 15]
```diff
@@ -137,6 +138,18 @@
...
+     * We can add streams to be ignored during the
+     * reconstruction of the state (e.g. raw streams)
+     */
+    @Getter
+    private Set<UUID> streamsToIgnore = new HashSet<>();
+
+    /**
+     * We can register streams with non-default type
+     */
+    private final Map<UUID, Builder> customTypeStreams = new HashMap<>();
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651447)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java  4
         

Clones removed
==============
+ test/src/test/java/org/corfudb/runtime/collections/SMRMapTest.java  -1
+ test/src/test/java/org/corfudb/runtime/collections/CorfuMapTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653419&bid=15584306)

**Comment by annym** [comment]
LGTM, we need to rebase. Also, not sure why the IF failed for two tests, we might want to run it a couple of times to be sure.

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/SMRObject.java  4
         

Clones removed
==============
+ test/src/test/java/org/corfudb/runtime/collections/SMRMapTest.java  -1
+ test/src/test/java/org/corfudb/runtime/collections/CorfuMapTest.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653419&bid=15584306)
